### PR TITLE
feat: verification of L1Batch witness (BFT-471)

### PR DESCRIPTION
**Pull request by pompon0**
Verification of an L1 batch, based on StoredBatchInfo. We extract the hash of the last block of the transaction via merkle path from the root_state (for batch n and n-1) and we recompute the rolling block hash based on transactions in the payload. The verification is not perfect as there are still some fields in payload for which we don't have a commitment - we should address that later.

Verification will be used for syncing L1Batches over p2p network (not implemented yet).

Also removed serde serialization from TransactionRequest which is unused - we don't need to maintain encoding compatibility for it.

### CODE REVIEW

File: core/node/consensus/src/batch.rs
Comment (StanislavBreadless) [inline core/node/consensus/src/batch.rs:246]: do we also want to add `anyhow::ensure` for the intermediate results of hashes, etc? It does not give any additional security, but allows to fail fast

File: core/node/consensus/src/tests.rs
Comment (StanislavBreadless) [inline core/node/consensus/src/tests.rs:522]: will it be resolved within this PR?

File: core/node/consensus/src/batch.rs
Comment (pompon0) [reply core/node/consensus/src/batch.rs:20]: This is what is stored on L1, so hash of the StoredBatchInfo, which contains a root_hash of the L2 state, which contains state of the SystemContext contract, which contains the L2 block hash.

File: core/node/consensus/src/tests.rs
Comment (pompon0) [reply core/node/consensus/src/tests.rs:522]: no, but will be resolved before we start using this logic. I don't have time right now to work on it.

File: core/node/consensus/src/batch.rs
Comment (pompon0) [reply core/node/consensus/src/batch.rs:246]: sg

File: core/node/consensus/src/batch.rs
Comment (StanislavBreadless) [reply core/node/consensus/src/batch.rs:20]: I would probably add a comment about it, from the outside it looks a bit confusing that the commitment for an L2 block is the batch hash

File: core/lib/dal/src/sync_dal.rs
Comment (popzxc) [inline core/lib/dal/src/sync_dal.rs:59]: I'm not an expert in SQL, but AFAIU `BETWEEN` operator is not ordered (e.g. arguments can come in either order), and so it looks like this query will work incorrectly if called with an empty range. Maybe bail if `numbers.is_empty()`?

File: core/lib/dal/src/transactions_web3_dal.rs
Comment (popzxc) [inline core/lib/dal/src/transactions_web3_dal.rs:407]: Same

File: core/node/consensus/src/batch.rs
Comment (popzxc) [inline core/node/consensus/src/batch.rs:168]: Nit: Maybe add `context` explaining which proof verification failed exactly? Would probably speed up debugging if ever occurs in a real env.

File: core/node/consensus/Cargo.toml
Comment (popzxc) [inline core/node/consensus/Cargo.toml:44]: Same

File: core/node/consensus/src/batch.rs
Comment (popzxc) [inline core/node/consensus/src/batch.rs:52]: Bikeshedding: maybe `system_context_addr`? I find `LastBlockWitness::addr()` a bit confusing.

**Review by popzxc — COMMENTED** [review]
Generally looks good, a few minor things.

File: core/lib/types/src/l2/mod.rs
Comment (pompon0) [reply core/lib/types/src/l2/mod.rs:212]: not exactly, because we need to call `req.get_signed_bytes(&sig)` afterward anyway

File: core/lib/dal/src/sync_dal.rs
Comment (pompon0) [reply core/lib/dal/src/sync_dal.rs:59]: thx, good catch

File: core/node/consensus/src/batch.rs
Comment (pompon0) [reply core/node/consensus/src/batch.rs:168]: done

File: core/node/consensus/src/batch.rs
Comment (pompon0) [reply core/node/consensus/src/batch.rs:52]: done

**Review by popzxc — APPROVED** [review]
LGTM, but would probably still suggest to wait for an approval from Stas

**Review by StanislavBreadless — APPROVED** [review]

**Pull request merged**
